electrum/lnrouter.py

```diff
@@ -161,12 +167,229 @@ def is_fee_sane(fee_msat: int, *, payment_amount_msat: int) -> bool:
+        return hint
+
+    @with_lock
+    def update_can_send(self, node_from: bytes, node_to: bytes, channel_id: ShortChannelID, amount: int):
+        hint = self.get_hint(channel_id)
+        hint.update_can_send(node_from < node_to, amount)
+
+    @with_lock
+    def update_cannot_send(self, node_from: bytes, node_to: bytes, channel_id: ShortChannelID, amount: int):
+        hint = self.get_hint(channel_id)
+        hint.update_cannot_send(node_from < node_to, amount)
+
+    def penalty(self, node_from: bytes, node_to: bytes, channel_id: ShortChannelID, amount: int) -> float:
+        """Gives a penalty when sending from node1 to node2 over channel_id with an
+        amount in units of millisatoshi.
+
+        The penalty depends on the can_send and cannot_send values that was
+        possibly recorded in previous payment attempts.
+
+        A channel that can send an amount is assigned a penalty of zero, a
+        channel that cannot send an amount is assigned an infinite penalty.
+        If the sending amount lies between can_send and cannot_send, there's
+        uncertainty and we give a default penalty. The default penalty
+        serves the function of giving a positive offset (the Dijkstra
+        algorithm doesn't work with negative weights), from which we can discount
+        from. There is a competition between low-fee channels and channels where
+        we know with some certainty that they can support a payment. The penalty
+        ultimately boils down to: how much more fees do we want to pay for
+        certainty of payment success? This can be tuned via DEFAULT_PENALTY_BASE_MSAT
+        and DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH. A base _and_ relative penalty
+        was chosen such that the penalty will be able to compete with the regular
+        base and relative fees.
+        """
+        # we only evaluate hints here, so use dict get (to not create many hints with self.get_hint)
+        hint = self._liquidity_hints.get(channel_id)
+        if not hint:
+            can_send, cannot_send = None, None
+        else:
+            can_send = hint.can_send(node_from < node_to)
+            cannot_send = hint.cannot_send(node_from < node_to)
+
+        if cannot_send is not None and amount >= cannot_send:
+            return inf
+        if can_send is not None and amount <= can_send:
+            return 0
+        return fee_for_edge_msat(amount, DEFAULT_PENALTY_BASE_MSAT, DEFAULT_PENALTY_PROPORTIONAL_MILLIONTH)
+
+    @with_lock
+    def add_to_blacklist(self, node_from: bytes, node_to: bytes, channel_id: ShortChannelID):
+        hint = self.get_hint(channel_id)
+        hint.is_blacklisted = True
+        now = int(time.time())
+        hint.timestamp = now
+
+    @with_lock
+    def get_blacklist(self) -> Set[ShortChannelID]:
+        now = int(time.time())
+        return set(k for k, v in self._liquidity_hints.items() if now - v.timestamp < BLACKLIST_DURATION)
```

Comment:
This is a bit confusing. So `hint.is_blacklisted` is not used?
Could we not merge `.is_blacklisted` and `.timestamp` to a single thing somehow?
Maybe just have a `.blacklist_timestamp` field, initialised to 0, and deduce whether a the edge is blacklisted from that and the `BLACKLIST_DURATION` global.

Reply:
Oh yes, somehow this slipped through. I changed to use only the timestamp.